homeassistant/components/binary_sensor/bmw_connected_drive.py

```diff
@@ -0,0 +1,123 @@
+"""
+Reads vehicle status from BMW connected drive portal.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.bmw_connected_drive/
+"""
+import logging
+import asyncio
+
+from homeassistant.components.bmw_connected_drive import DOMAIN as BMW_DOMAIN
+from homeassistant.components.binary_sensor import BinarySensorDevice
+
+DEPENDENCIES = ['bmw_connected_drive']
+
+_LOGGER = logging.getLogger(__name__)
+
+SENSOR_TYPES = {
+    'all_lids_closed': ['Doors', 'opening'],
+    'all_windows_closed': ['Windows', 'opening'],
+    'door_lock_state': ['Door lock state', 'safety']
+}
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the BMW sensors."""
+    accounts = hass.data[BMW_DOMAIN]
+    _LOGGER.debug('Found BMW accounts: %s',
+                  ', '.join([a.name for a in accounts]))
+    devices = []
+    for account in accounts:
+        for vehicle in account.account.vehicles:
+            for key, value in sorted(SENSOR_TYPES.items()):
+                device = BMWConnectedDriveSensor(account, vehicle, key,
+                                                 value[0], value[1])
+                devices.append(device)
+    add_devices(devices)
+
+
+class BMWConnectedDriveSensor(BinarySensorDevice):
+    """Representation of a BMW vehicle binary sensor."""
+
+    def __init__(self, account, vehicle, attribute: str, sensor_name,
+                 device_class):
+        """Constructor."""
+        self._account = account
+        self._vehicle = vehicle
+        self._attribute = attribute
+        self._name = sensor_name
+        self._device_class = device_class
+        self._state = None
+        self._unit_of_measurement = None
+
+    @property
+    def should_poll(self) -> bool:
+        """Data update is triggered from BMWConnectedDriveEntity."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the binary sensor."""
+        return self._name
+
+    @property
+    def device_class(self):
+        """Return the class of the binary sensor."""
+        return self._device_class
+
+    @property
+    def is_on(self):
+        """Return the state of the binary sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Get the unit of measurement."""
+        return self._unit_of_measurement
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the binary sensor."""
+        vehicle_state = self._vehicle.state
+        result = {
+            'last_update': vehicle_state.timestamp,
```

Comment:
Time stamp is not allowed in state attributes. The state should not update just because of a time change.

Reply:
This shows the last update time which is received from the car via the bmw_connected_drive module.
This time stamp is only updated together with other updates received from the car, e.g. an update on the mileage, see the screenshot below.

It's the same usage as in the [Mercedes ME](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/binary_sensor/mercedesme.py#L57) component.

![screenshot_7](https://user-images.githubusercontent.com/11230573/36565528-003e73c6-1821-11e8-87f9-09e7fa7f66c4.png)

Reply:
@gerard33 
I get your point, I agree it's quite useful to see when the last measurement was taken. But isn't the information redundant? You would also get the age of the information from the headline "3 hours ago".

Reply:
Yes, this is redundant, please remove it.

Reply:
As a matter of fact it is not redudant. The last update contains the time on which the update from the car is received, but that doesn't always mean that every sensor is being updated. So this can result in a different time shown at the time information in the headline than at the last update time. See the example with the below screenshots.
With this update the door lock state has changed, but the windows state hasn't, resulting in different times for the windows sensor in the headline versus the last update time.

I think it has added value to show this attribute as you can easily see if the sensor has changed or not as part of the last update received from the car.

![img_8122_2](https://user-images.githubusercontent.com/11230573/37354725-9a629782-26e2-11e8-89d8-547cd253aecd.png)
![img_8123_2](https://user-images.githubusercontent.com/11230573/37354726-9a7c0ae6-26e2-11e8-93fd-589ae6637a9f.png)

Reply:
If a state attribute changes, the last_updated attribute of the state will be updated with the time, so this information is already available in the the state machine. If a user wants to know about this we offer the template sensor which can display state attributes including last_updated. We shouldn't duplicate this information in the state machine.